vm/src/vm.rs

```diff
@@ -889,20 +884,25 @@ impl VirtualMachine {
 
     fn _invoke(&self, callable: &PyObjectRef, args: PyFuncArgs) -> PyResult {
         vm_trace!("Invoke: {:?} {:?}", callable, args);
-        if let Some(slot_call) = callable.lease_class().slots.call.as_ref() {
-            self.trace_event(TraceEvent::Call)?;
-            let args = args.insert(callable.clone());
-            let result = slot_call(self, args);
-            self.trace_event(TraceEvent::Return)?;
-            result
-        } else if callable.lease_class().has_attr("__call__") {
-            self.call_method(&callable, "__call__", args)
-        } else {
+        let invoked = callable.class().first_in_mro(|cls| {
+            cls.slots.call.load().map(|slot_call| {
+                // NOTE: unnessessary copy
+                let args = PyFuncArgs {
+                    args: args.args.clone(),
+                    kwargs: args.kwargs.clone(),
+                };
+                self.trace_event(TraceEvent::Call)?;
+                let result = slot_call(callable.clone(), args, self);
+                self.trace_event(TraceEvent::Return)?;
+                result
+            })
+        });
+        invoked.unwrap_or_else(|| {
```

Comment:
You could just have `first_in_mro(|cls| cls.slots.call.load())` and invoke it outside of the closure.